decisions/0004-generic-data-storage.md

```diff
@@ -0,0 +1,80 @@
+- Reduction of code bloat, particularly for POD classes.
+- Removal of the custom `List` and `ListItem` classes, upon which `GenericList` is currently based.
+
+## Considered Options
+
+Assuming that the underlying `List` is to be replaced by a `std::map<std::string, X>` where the `std::string` is the name of the data, the heterogeneous data item `X` can be represented as:
+
+1. `std::variant`
+2. `std::any`
+3. `boost::fusion`
+
+## Decision Outcome
+
+Chosen option: "`std::any`", because it requires no additional dependencies (unlike `boost::fusion`) and has a smaller memory footprint per-item (unlike `std::variant`). By registering templated `std::function`s specific to individual class types, the necessary functionality can provided on a per-class basis, either calling class functions or implementing the necessary code in the function body. This also allows good separation between the `GenericList` and the various operations - serialisation, deserialisation, etc. Since the desired `std::map` is to contain values that are heterogenous (or are, for instance, a `std::tuple` containing a member which is the heterogeneous value) the benefits of `boost::fusion` do not seem readily applicable.
+
+### Positive Consequences
+
+- Facile implementation of new classes of functors acting upon specific class types.
+- No inheritance required
+- Significant code reduction
+
+### Negative Consequences
+
+- The use of registered function pointers rather than a series of templated specialisations (as would be the case with `std::variant`) is perhaps less transparent, however judicious use of `typedef`s helps to retain readability of the code.
+
+## Pros and Cons of the Options
+
+### `std::variant`
+
+- Good, because requires no additional dependencies
+- Good, because permits easy traversal of all data in the `std::map` via the use of a type matching visitor or overloaded operators
+- Bad, because requires all functions for all types to be defined within the templated `GenericList` function, potentially leading to executable bloat and slowdown from extensive header inclusion from the base `GenericList` class
+- Bad, because any item takes on the memory footprint of the largest class in the `std::variant` (e.g. `bool` vs `MyLargeClass`)
+
+### `std::any`
+
+- Good, because requires no additional dependencies
+- Good, because the size of the `std::any` is defined by the individual type stored
+- Bad, because there is no equivalent of `std::visit`, so (e.g.) templated function registrations must be employed to provide functionality
+- Good, because the use of templated function registrations can be made outside of the templated `GenericItem` class functions, removing direct include dependencies in the build chain
+- Bad, because need to indicate contained types at runtime via `typeid`
```

Comment:
```suggestion
- Bad, because need to indicate contained types at runtime via `typeid`.  This could lead to type errors that aren't caught be the compiler, though this should only occur if someone tries to do something "clever".
```